File: kinotic-github/src/main/java/org/kinotic/os/github/api/services/ProjectGitHubRepoService.java
package org.kinotic.os.github.api.services;

import org.kinotic.core.api.annotations.Publish;
import org.kinotic.core.api.crud.IdentifiableCrudService;
import org.kinotic.os.github.api.model.AvailableRepo;
import org.kinotic.os.github.api.model.ProjectGitHubRepoLink;

import java.util.List;
import java.util.concurrent.CompletableFuture;

/**
 * Service the project-settings UI uses to read and mutate the link between a Kinotic
 * Project and a GitHub repository. Links are 1:1 — at most one repo per project.
 */
@Publish
public interface ProjectGitHubRepoService extends IdentifiableCrudService<ProjectGitHubRepoLink, String> {

    /**
     * Lists the repositories visible under the caller's organization's installation,
     * for populating the repo dropdown in the link UI. Calls GitHub on every invocation;
     * not cached because the org admin may have just toggled repo access.
     */
    CompletableFuture<List<AvailableRepo>> listAvailableRepos();

    /**
     * Creates or replaces the link for the given project. Validates that the chosen
     * {@code repoFullName} is reachable through the org's installation before persisting.
     */
    CompletableFuture<ProjectGitHubRepoLink> linkProject(String projectId, String repoFullName);

    /** Removes the link for the given project, if any. */
    CompletableFuture<Void> unlinkProject(String projectId);

    /**
     * Returns the link for the given project, or {@code null} when none exists.
     */
    CompletableFuture<ProjectGitHubRepoLink> findByProject(String projectId);
}

File: kinotic-frontend/src/services/ProjectGitHubRepoService.ts
import {Kinotic, type IServiceProxy} from '@kinotic-ai/core'

export interface AvailableRepo {
    repoId: string
    fullName: string
    defaultBranch: string
    privateRepo: boolean
}

export interface ProjectGitHubRepoLink {
    id: string
    projectId: string
    organizationId: string
    installationId: string
    repoFullName: string
    repoId: string
    defaultBranch: string
    updated?: string
}

/**
 * Service proxy for the published org-scoped ProjectGitHubRepoService. Drives the
 * project-settings repo dropdown and link/unlink actions.
 */
export class ProjectGitHubRepoService {

    private serviceProxy: IServiceProxy

    constructor() {
        this.serviceProxy = Kinotic.serviceProxy(
            'org.kinotic.os.github.api.services.ProjectGitHubRepoService')
    }

    public listAvailableRepos(): Promise<AvailableRepo[]> {
        return this.serviceProxy.invoke('listAvailableRepos', [])
    }

    public linkProject(projectId: string, repoFullName: string): Promise<ProjectGitHubRepoLink> {
        return this.serviceProxy.invoke('linkProject', [projectId, repoFullName])
    }

    public unlinkProject(projectId: string): Promise<void> {
        return this.serviceProxy.invoke('unlinkProject', [projectId])
    }

    public findByProject(projectId: string): Promise<ProjectGitHubRepoLink | null> {
        return this.serviceProxy.invoke('findByProject', [projectId])
    }
}

export const PROJECT_GITHUB_REPO_SERVICE = new ProjectGitHubRepoService()
